[core] Fix GCS client retry

vitsai opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

Recreate the grpc channel during retries, in case the originally created channel was faulty.

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

ray start was seeing a 3x slowdown due to failing reconnects on a faulty channel.

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

https://github.com/ray-project/ray/issues/38509

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

iycheng reviewed (commented):
I think it should be fine, but @rkooo567 @jjyao  could you give it another look?

jjyao commented:
What's the before and after time for `ray start --head`?

jjyao commented on python/ray/_private/node.py:
```diff
@@ -1238,6 +1235,9 @@ def start_head_processes(self):
             raise_on_api_server_failure = self._ray_params.include_dashboard
 
+        assert self.get_gcs_client() is not None
+        self._write_cluster_info_to_kv()
```
I think it's fragile: there is a hidden assumption here that monitor and ray client server don't need to access cluster info populated by `_write_cluster_info_to_kv`

vitsai replied:
So I guess a more enduring fix might look something like: let GCS serve the kv store before the rest of the server has fully initialized, but there's the whole problem of GCS being single-threaded

vitsai replied:
passing the metadata to the processes when they're started is not the best solution, but it's /a/ solution

vitsai commented on python/ray/_private/node.py:
```diff
@@ -1052,7 +1052,6 @@ def start_gcs_server(self):
         # when possible.
         self._gcs_address = f"{self._node_ip_address}:" f"{gcs_server_port}"
         # Initialize gcs client, which also waits for GCS to start running.
-        self._init_gcs_client()
```
`get_gcs_client` will call `_init_gcs_client` the first time it's called

vitsai commented:
Time has gone back down from 5.09s to 1.05s

vitsai commented:
Moved the sync point back to where it was and fixed a bug in the retry logic. It's now 1.97 seconds, which is around where it was originally. Note that if we move the sync point to later in time as the earlier revision did, we can further optimize to ~1.05 at the expense of brittleness.

jjyao commented on python/ray/_private/node.py:
```diff
@@ -1133,6 +1132,7 @@ def start_monitor(self):
         process_info = ray._private.services.start_monitor(
             self.gcs_address,
             self._logs_dir,
+            self._session_name,
```
These are no longer needed

vitsai replied:
removed

jjyao reviewed (commented):
I like the new fix :)

jjyao commented on src/ray/gcs/gcs_client/gcs_client.cc:
```diff
@@ -191,6 +186,9 @@ Status PythonGcsClient::Connect(const ClusterID &cluster_id,
         return HandleGcsError(reply.status());
       }
       std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+      channel_ =
+          rpc::GcsRpcClient::CreateGcsChannel(options_.gcs_address_, options_.gcs_port_);
+      node_info_stub_ = rpc::NodeInfoGcsService::NewStub(channel_);
```
What's the root cause?

vitsai replied:
The channel was not being recreated during retries, so an RPC that failed the first time was doomed to fail for every subsequent retry (spaced 1s apart) since a faulty channel would be returned. We also don't check whether the channel is faulty when we first create it.

vitsai replied:
removed

rkooo567 replied:
Can we just use https://github.com/ray-project/ray/pull/39017 (cc @rynewang)?

vitsai replied:
My understanding is that the linked PR is for connection fluctuations on a working channel, whereas in this case, it's not clear that the channel created is working at all (so it is necessary to create a new one). i.e. RPCs fail on the bad channel for 5 seconds, whereas the server comes up within a second and a half -- so wait_for_ready will either be slower than recreating, or it will never work.

rkooo567 replied:
I see, that makes sense. So this is the case retry is better than waiting until channel is recovered from a bad state. 

Bad channel -> is it because the channel may never recover because of GCS restart?

vitsai replied:
Not sure whether the state on the channel is laggy (by several seconds), or if the created channel is just permanently bad (maybe because GCS was not available at time of creation, although in this case it's still starting up, not restarting)

rkooo567 replied:
So the laggy case could be handled by wait_for_ready (the transient state should be recovered), but the closed channel (without proper shutdown) cannot be handled by this IIUC? Anyway it makes sense to retry here.

Btw, @rynewang don't we still need wait_for_ready here still at least for a few seconds in case it takes time to establish the channel? It seems like this approach won't work if the channel connection cannot be estabilshed within 1 second?

rynewang replied:
This confuses me a bit. My understanding is a channel can always be created, albeit in a TRANSIENT_FAILURE state if remote is not accessible, and will retry to connect if in that state according to the [grpc doc](https://grpc.github.io/grpc/cpp/md_doc_connectivity-semantics-and-api.html). What does it mean by `since a faulty channel would be returned` and `doomed to fail for every subsequent retry`? I guess with [a wait for retry PR](https://github.com/ray-project/ray/pull/39017) it should work.

rkooo567 replied:
This makes sense to me. Btw, what's the reason why the first channel is not ready in the first place? Is it because we try connecting to GCS before it is ready? (and exponential backoff will make it longer time to make transient failure -> ready state?)

rkooo567 commented on python/ray/scripts/scripts.py:
```diff
@@ -938,6 +939,8 @@ def start(
     assert ray_params.gcs_address is not None
     ray._private.utils.write_ray_address(ray_params.gcs_address, temp_dir)
 
+    logger.info("Started Ray in: " + str(datetime.now() - start_time) + " seconds.")
```
We should remove this before we merge it. Or we can add the second information around the starting log that we already have

vitsai replied:
Moved it to the starting log.

rkooo567 replied:
I don't see the change yet?

vitsai replied:
sorry, pushed after responding, should be there now

rkooo567 commented on src/ray/gcs/gcs_client/gcs_client.cc:
```diff
@@ -199,6 +197,11 @@ Status PythonGcsClient::Connect(const ClusterID &cluster_id,
   }
 
   RAY_CHECK(!cluster_id_.IsNil()) << "Unexpected nil cluster ID.";
+  kv_stub_ = rpc::InternalKVGcsService::NewStub(channel_);
```
why do we not reset things like `node_info_stub_`?

vitsai replied:
We initialize `node_info_stub_` at the time of creating the channel, and reset it within the loop, so by this point it should already be initialized. The remaining stubs here have not been initialized, so we do that here.

rkooo567 commented:
Please rewrite the PR title & description!

rynewang commented on src/ray/gcs/gcs_server/gcs_server.h:
```diff
@@ -98,6 +97,9 @@ class GcsServer {
   bool IsStopped() const { return is_stopped_; }
 
+  /// Retrieve cluster ID
+  const ClusterID GetClusterId() const { return rpc_server_.GetClusterId(); }
```
const & ?

vitsai replied:
I'll change it, but with RVO it shouldn't matter very much

jjyao commented on python/ray/_private/node.py:
```diff
@@ -1052,7 +1052,6 @@ def start_gcs_server(self):
         # when possible.
         self._gcs_address = f"{self._node_ip_address}:" f"{gcs_server_port}"
         # Initialize gcs client, which also waits for GCS to start running.
-        self._init_gcs_client()
```
The above comment needs to be removed as well.

jjyao commented on python/ray/scripts/scripts.py:
```diff
@@ -769,7 +770,10 @@ def start(
         # are still in place
         cli_logger.newline()
-        startup_msg = "Ray runtime started."
+        startup_msg = (
+            "Ray runtime started "
+            f" ({(datetime.now() - start_time).total_seconds()} seconds)."
```
Don't change the public message

rkooo567 reviewed (changes requested):
LGTM. Before we merge, Should we add some sort of test to measure ray start latency? To catch issues like this in the future? (I will approve it once it is resolved).

vitsai commented:
The value is pretty platform dependent; 6s -> 2s was on some AWS machine, but on Mac it is 10.7s -> 6.7s. Seems like a test here would just be flaky and also not add much value since it would have to test the ceiling of all the start times. I guess we could choose one specific platform + machine to always run the test on?

rkooo567 commented:
@vitsai let's remove it from CI test and add it to microbenchmark. Let's do it in a follow up PR

rkooo567 commented:
premrege failure

rkooo567 commented:
Will merge once it passes https://buildkite.com/ray-project/release-tests-pr/builds/51818

rkooo567 commented:
Passes
```


[INFO 2023-09-01 06:01:41,857] anyscale_job_runner.py: 145  Output: {'collected_metrics': True, 'last_prepare_time_taken': 6.347128261999956, 'prepare_return_codes': [0], 'return_code': 0, 'total_time_taken': 360.87290291500005, 'uploaded_artifact': False, 'uploaded_metrics': True, 'uploaded_results': True, 'workload_time_taken': 307.565308664}
--
  |  

<br class="Apple-interchange-newline">
```

rickyyx commented:
I think this also fixes https://github.com/ray-project/ray/issues/39111 

I ran the test 100 times on this PR, and it now only failed 4 out of 100 times: https://buildkite.com/ray-project/oss-ci-build-pr/builds/34672#018a4def-7346-4291-8285-34cbaa818016
<img width="882" alt="image" src="https://github.com/ray-project/ray/assets/11676094/b3c464a4-1d2f-4c74-b661-49c199563346">
